uobtheatre/bookings/models.py

```diff
@@ -521,36 +610,107 @@ def get_ticket_diff(
+            raise BookingTransferToDifferentProductionException
 
-        # If a payment is created set the booking as paid
-        if payment.status == Transaction.Status.COMPLETED:
-            self.complete(payment)
+        if not performance.is_bookable:
+            raise NotBookableException
 
-        return payment
+        if performance.capacity_remaining < self.tickets.count():
+            raise CapacityException(
+                f"There are only {performance.capacity_remaining} seats remaining for the selected performance"
+            )
 
-    def complete(self, payment: Transaction = None):
+    def _check_transfer_booking(self):
+        # If the booking is already completed. Note this also convers if it has
+        # already been transfered as once a transfer in completed the original
+        # booking is cancelled.
+        if self.status != Booking.Status.PAID:
+            raise BookingTransferBookingNotPaidException(self.get_status_display())
+
+    def create_transfer(self, performance: "Performance") -> "Booking":
         """
-        Complete the booking (after it has been paid for) and send the
-        confirmation email.
+        Transfer the booking to a different performance.
+
+        A successful transfer creates a new booking which copies the attributes
+        from the original. The original is cancelled and the transfered to
+        attribute is assigned as the new booking.
         """
-        self.status = Payable.Status.PAID
-        self.save()
-        self.send_confirmation_email(payment)
+
+        self._check_transfer_booking()
+        self._check_transfer_performance(performance)
+
+        # This will delete any exisiting IN_PROGRESS booking that the user has
+        # for this performance (this includes transfers)
+        delete_user_drafts(self.user, performance.id)
+
+        # Create a booking transfer model
+        new_booking = self.clone()
+        new_booking.status = Payable.Status.IN_PROGRESS
+        new_booking.performance = performance
+        new_booking.transfered_from = self
+        new_booking.save()
+
+        # Copy across all tickets which can be copied
+        for ticket in self.tickets.all():
+            # TODO maybe define can_transfer_to on ticket model
+            # If there is capcity for this ticket in the other performance then
+            # copy this ticket to the new booking
+            try:
+                performance.check_capacity(ticket.qs)
+
+                # Clone ticket to new booking
+                new_ticket = ticket.clone()
+                new_ticket.booking = new_booking
+                new_ticket.save()
+            except (
+                UnassignedSeatGroupException,
+                UnassignedConcessionTypeException,
+                CapacityException,
+            ) as exc:
+                print(
+                    f"Not copying accross ticket {ticket.seat_group} as {exc.__class__}"
+                )
```

Comment:
I assume this is going to be replaced with something useful?

Reply:
I believe this is all part of the plan... If tickets cannot be copied across, they just won't be and the users will have to select new tickets

Reply:
Yeah but it should be pretty clear in the UI that not all the tickets have been copied across as they will be presented with the standard (ish) booking screen. If you want to get really fancy on the UI side you can check if the length of tickets has changed and display a warning. I dont think an exception is right for this as it expected behaviour... Is there a good way to do warning?

Reply:
It's a bit grim, but I think this function should return a tuple or something, including the newly created booking and a list of warnings. These should then be passed through in the GQL schema. We shouldn't have print commands running on production.